### PR TITLE
feat(backend/frontend/shared): pagination to the add course pop-up window tm-537

### PR DESCRIPTION
**Pull request by Ann-Maksay**
UPD: The search returns 10 items, not 12.

- Course uploading

search-load-more

- Changing the value in the search bar

search-load-more-new-search

- Close and reopen the modal window

search-load-more-new-window

- Adding a course makes the button disabled in all lists

search-load-more-recomended

UPD 2: button
image

### CODE REVIEW

File: apps/frontend/src/pages/overview/libs/components/add-course-modal/add-course-modal.tsx [hunk 1]
```diff
@@ -153,22 +183,31 @@ const AddCourseModal: React.FC<Properties> = ({
...
-											courses={recommendedCourses}
-											onAddCourse={handleAddCourse}
+									<>
+										<Button
+											className={styles["load-more-button"]}
+											isDisabled={isLoadMore}
+											isLoading={isLoadMore}
+											label="Load more"
+											onClick={handleLoadMore}
 										/>
```
Comment (YourTeaChan) [inline apps/frontend/src/pages/overview/libs/components/add-course-modal/add-course-modal.tsx:199]: I think a small button would look better

File: apps/frontend/src/libs/components/course/course.tsx [hunk 2]
```diff
@@ -31,9 +37,11 @@ const Course: React.FC<Properties> = ({
...
 	const isLoading =
 		addedVendorCourseDataStatuses[vendorCourseId] === DataStatus.PENDING;
-	const isDisabled =
-		isLoading ||
-		addedVendorCourseDataStatuses[vendorCourseId] === DataStatus.FULFILLED;
+
+	const isUserHasCourse = checkIsSearchedCourse(course)
+		? course.isUserHasCourse
+		: true;
+	const isDisabled = isLoading || isUserHasCourse;
```
Comment (Vfivew) [inline apps/frontend/src/libs/components/course/course.tsx:44]: A1

**Review by Vfivew — CHANGES_REQUESTED** [review]

File: packages/shared/src/modules/courses/libs/types/courses-search-response-dto.type.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { type CourseSearchResponseDto } from "./course-search-response-dto.type.js";
+
+type CoursesSearchResponseDto = {
```
Comment (YourTeaChan) [inline packages/shared/src/modules/courses/libs/types/courses-search-response-dto.type.ts:3]:
```suggestion
type CourseSearchGetAllResponseDto = {
```

File: apps/frontend/src/libs/components/course/course.tsx (hunk 2 above)
Comment (Ann-Maksay) [reply apps/frontend/src/libs/components/course/course.tsx:44]: It is not a [Nullish coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescingl). This checks whether a isLoading is true or false, not whether it exists.

Comment (Vfivew) [reply apps/frontend/src/libs/components/course/course.tsx:44]: My mistake. Sorry

**Review by Vfivew — APPROVED** [review]

**Review by YourTeaChan — APPROVED** [review]

File: apps/frontend/src/modules/courses/slices/courses.slice.ts [hunk 4]
```diff
@@ -68,6 +74,18 @@ const { actions, name, reducer } = createSlice({
 		builder.addCase(userCoursesActions.add.fulfilled, (state, action) => {
 			const { vendorCourseId } = action.meta.arg;
 
+			state.searchedCourses = state.searchedCourses.map((course) => {
+				return course.vendorCourseId === vendorCourseId
+					? { ...course, isUserHasCourse: true }
+					: course;
```
Comment (Oleksandra-Masiuk) [inline apps/frontend/src/modules/courses/slices/courses.slice.ts:80]:
```suggestion
				const hasUserCourse = course.vendorCourseId === vendorCourseId;
				return { ...course, hasUserCourse };
```

File: apps/frontend/src/modules/courses/slices/courses.slice.ts [hunk 5]
```diff
@@ -68,6 +74,18 @@ const { actions, name, reducer } = createSlice({
...
+			});
+
+			state.recommendedCourses = state.recommendedCourses.map((course) => {
+				return course.vendorCourseId === vendorCourseId
+					? { ...course, isUserHasCourse: true }
+					: course;
```
Comment (Oleksandra-Masiuk) [inline apps/frontend/src/modules/courses/slices/courses.slice.ts:86]:
```suggestion
			state.recommendedCourses = state.recommendedCourses.map((course) => {
				const hasUserCourse = course.vendorCourseId === vendorCourseId;
				return { ...course, hasUserCourse };
```

File: apps/frontend/src/libs/components/course/libs/helpers/check-is-searched-course.helper.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import {
+	type CourseDto,
+	type CourseSearchResponseDto,
+} from "~/modules/courses/courses.js";
+import { type UserCourseResponseDto } from "~/modules/user-courses/user-courses.js";
+
+const checkIsSearchedCourse = (
+	dto: CourseDto | CourseSearchResponseDto | UserCourseResponseDto,
+): dto is CourseSearchResponseDto => {
+	return "isUserHasCourse" in dto;
```
Comment (Oleksandra-Masiuk) [inline apps/frontend/src/libs/components/course/libs/helpers/check-is-searched-course.helper.ts:10]:
```suggestion
	return "hasUserCourse" in dto;
```

**Review by Oleksandra-Masiuk — CHANGES_REQUESTED** [review]

**Review by Sytnikovakatya — APPROVED** [review]

File: apps/frontend/src/modules/courses/slices/courses.slice.ts (hunk 4 above)
Comment (Ann-Maksay) [reply apps/frontend/src/modules/courses/slices/courses.slice.ts]: We want to change hasUserCourse only in the target component, and this code will cause this value to become false in all other components besides it

Comment (Oleksandra-Masiuk) [reply apps/frontend/src/modules/courses/slices/courses.slice.ts]: In this case, can you do it like this?
```
const hasUserCourse = course.vendorCourseId === vendorCourseId;
return hasUserCourse
         ? { ...course, hasUserCourse }
	 : course;```

File: apps/frontend/src/libs/hooks/use-app-form/use-app-form.hook.ts [hunk 7]
```diff
@@ -27,6 +28,7 @@ type ReturnValue<T extends FieldValues = FieldValues> = {
 	control: Control<T, null>;
 	errors: FieldErrors<T>;
 	formState: UseFormStateReturn<T>;
+	getValues: UseFormGetValues<T>;
```
Comment (what1s1ove) [inline apps/frontend/src/libs/hooks/use-app-form/use-app-form.hook.ts:31]:
```suggestion
```
you need to use onChange mode to get values

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/hooks/use-app-form/use-app-form.hook.ts (hunk 7 above)
Comment (Ann-Maksay) [reply apps/frontend/src/libs/hooks/use-app-form/use-app-form.hook.ts:31]: @what1s1ove 
I don't understand how to do this and why I can't use getValues.
I need it to load more courses for an existing search query. If onChange is triggered, there is a different logic. 
The only thing that comes to my mind is to write the value somewhere in the state onChange but that seems redundant.

Comment (what1s1ove) [reply apps/frontend/src/libs/hooks/use-app-form/use-app-form.hook.ts:31]: Got the problem. Merged

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**